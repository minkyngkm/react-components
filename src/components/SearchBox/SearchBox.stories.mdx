import { ArgsTable, Canvas, Meta, Story } from "@storybook/addon-docs/blocks";

import SearchBox from "./SearchBox";

<Meta title="SearchBox" component={SearchBox} />

export const Template = (args) => <SearchBox {...args} />;

### SearchBox

This is a [React](https://reactjs.org/) component for the Vanilla [SearchBox](https://docs.vanillaframework.io/patterns/search-box/).

Search boxes enable search functionality on a page and are typically used in a navigation bar. SearchBox elements should be wrapped within a form element that handles its own submit action.

### Props

<ArgsTable of={SearchBox} />

### Default

<Canvas>
  <Story name="Default">{Template.bind({})}</Story>
</Canvas>

### Disabled

<Canvas>
  <Story name="Disabled">
    <SearchBox disabled />
  </Story>
</Canvas>

### External state

If you wish to control the value of the input via external state you can set
the `externallyControlled` prop and provide an `onChange` method to update the
state and the `value` from state.

<Canvas>
  <Story name="External state">
    <SearchBox externallyControlled onChange={() => {}} value="..." />
  </Story>
</Canvas>

### Navigation

<Canvas>
  <Story name="Navigation">
    <header id="navigation" className="p-navigation">
      <div className="p-navigation__row--full-width">
        <div className="p-navigation__banner">
          <div className="p-navigation__logo">
            <a className="p-navigation__item" href="#">
              <img
                className="p-navigation__image"
                src="https://assets.ubuntu.com/v1/5d6da5c4-logo-canonical-aubergine.svg"
                alt=""
                width="95"
              />
            </a>
          </div>
          <a
            href="#navigation"
            className="p-navigation__toggle--open"
            title="menu"
          >
            Menu
          </a>
          <a
            href="#navigation-closed"
            className="p-navigation__toggle--close"
            title="close menu"
          >
            Close menu
          </a>
        </div>
        <nav className="p-navigation__nav">
          <span className="u-off-screen">
            <a href="#main-content">Jump to main content</a>
          </span>
          <ul className="p-navigation__items" role="menu">
            <li className="p-navigation__item" role="menuitem">
              <a href="#" class="p-navigation__link">
                Products
              </a>
            </li>
            <li className="p-navigation__item" role="menuitem">
              <a href="#" class="p-navigation__link">
                Services
              </a>
            </li>
            <li className="p-navigation__item" role="menuitem">
              <a href="#" class="p-navigation__link">
                Partners
              </a>
            </li>
            <li className="p-navigation__item" role="menuitem">
              <a href="#" class="p-navigation__link">
                About
              </a>
            </li>
            <li className="p-navigation__item" role="menuitem">
              <a href="#" class="p-navigation__link">
                Partners
              </a>
            </li>
          </ul>
          <SearchBox />
        </nav>
      </div>
    </header>
  </Story>
</Canvas>
