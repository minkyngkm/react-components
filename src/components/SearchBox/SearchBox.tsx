import classNames from "classnames";
import React from "react";

import Icon from "../Icon";

import type { ClassName } from "types";

export type Props = {
  /**
   * Whether autocomplete should be enabled for the search input.
   */
  autocomplete?: "on" | "off";
  /**
   * Optional classes to pass to the form element.
   */
  className?: ClassName;
  /**
   * Whether the input and buttons should be disabled.
   */
  disabled?: boolean;
  /**
   * Whether the input value will be controlled via external state.
   */
  externallyControlled?: boolean;
  /**
   * A function that will be called when the input value changes.
   */
  onChange?: (inputValue: string) => void;
  /**
   * A function that is called when the user clicks the search icon
   */
  onSearch?: () => void;
  /**
   * A search input placeholder message.
   */
  placeholder?: string;
  /**
   * The value of the search input when the state is externally controlled.
   */
  value?: string;
};

const SearchBox = ({
  autocomplete = "on",
  className,
  disabled,
  externallyControlled,
  onChange,
  onSearch,
  placeholder = "Search",
  value,
}: Props): JSX.Element => {
  const input = React.createRef<HTMLInputElement | null>();
  const resetInput = () => {
    onChange && onChange("");
    if (input.current) {
      input.current.value = "";
    }
  };

  const triggerSearch = () => {
    onSearch && onSearch();
  };

  return (
    <div className={classNames("p-search-box", className)}>
      <label className="u-off-screen" htmlFor="search">
        {placeholder || "Search"}
      </label>
      <input
        autoComplete={autocomplete}
        className="p-search-box__input"
        disabled={disabled}
        id="search"
        name="search"
        onChange={(evt) => onChange(evt.target.value)}
        placeholder={placeholder}
        ref={input}
        type="search"
        defaultValue={externallyControlled ? undefined : value}
        value={externallyControlled ? value : undefined}
      />
      {value && (
        <button
          className="p-search-box__reset"
          disabled={disabled}
          onClick={resetInput}
          type="reset"
        >
          <Icon name="close">Clear search field</Icon>
        </button>
      )}
      <button
        className="p-search-box__button"
        disabled={disabled}
        onClick={triggerSearch}
      >
        <Icon name="search">Search</Icon>
      </button>
    </div>
  );
};

SearchBox.displayName = "SearchBox";

export default SearchBox;
