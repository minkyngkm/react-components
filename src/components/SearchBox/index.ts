export { default } from "./SearchBox";
export type { Props as SearchBoxProps } from "./SearchBox";
