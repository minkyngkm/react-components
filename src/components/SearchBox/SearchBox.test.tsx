import { shallow, mount } from "enzyme";
import React from "react";

import SearchBox from "./SearchBox";

describe("SearchBox ", () => {
  it("renders", () => {
    const wrapper = shallow(<SearchBox />);
    expect(wrapper).toMatchSnapshot();
  });

  it("shows the clear button when there is a value", () => {
    const wrapper = shallow(<SearchBox value="admin" />);
    expect(wrapper.find(".p-search-box__reset").exists()).toBe(true);
  });

  it("can externally control the value", () => {
    const wrapper = shallow(
      <SearchBox externallyControlled onChange={jest.fn()} value="admin" />
    );
    let input = wrapper.find(".p-search-box__input");
    expect(input.prop("defaultValue")).toBe(undefined);
    expect(input.prop("value")).toBe("admin");
    wrapper.setProps({ value: "new-admin" });
    input = wrapper.find(".p-search-box__input");
    expect(input.prop("value")).toBe("new-admin");
  });

  it("should call onSearch prop", () => {
    const onSearchMock = jest.fn();
    const component = shallow(<SearchBox onSearch={onSearchMock} />);
    component.find(".p-search-box__button").simulate("click");
    expect(onSearchMock).toBeCalled();
  });

  it("should call onChange prop", () => {
    const onChangeMock = jest.fn();
    const event = {
      target: {
        value: "test",
      },
    };
    const component = shallow(<SearchBox onChange={onChangeMock} />);
    component.find(".p-search-box__input").simulate("change", event);
    expect(onChangeMock).toBeCalledWith(event.target.value);
  });

  it("can be found using the component name", () => {
    const WrappingComponent = () => (
      <div>
        <SearchBox />
      </div>
    );
    const wrapper = mount(<WrappingComponent />);
    expect(wrapper.find("SearchBox").exists()).toBe(true);
  });
});
