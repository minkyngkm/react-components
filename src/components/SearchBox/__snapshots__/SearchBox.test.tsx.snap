// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBox  renders 1`] = `
<div
  className="p-search-box"
>
  <label
    className="u-off-screen"
    htmlFor="search"
  >
    Search
  </label>
  <input
    autoComplete="on"
    className="p-search-box__input"
    id="search"
    name="search"
    onChange={[Function]}
    placeholder="Search"
    type="search"
  />
  <button
    className="p-search-box__button"
    onClick={[Function]}
  >
    <Icon
      name="search"
    >
      Search
    </Icon>
  </button>
</div>
`;
